### ISSUE TITLE
Wrong type inference when calling functions with side effects and strict null checks are enabled

### ISSUE THREAD

**Issue by kayahr** [Microsoft/TypeScript#19190]
**TypeScript Version:**  2.5.3 and 2.6.0-dev.20171015

**Code**

```ts
let text: string | null = null;

function setText(newText: string) {
    text = newText;
}

setText("Hello World");
if (text) {
    console.log(text.toLowerCase());
}

```

**Expected behavior:**

Code should compile with `--strictNullChecks`. Type of `text` should be `null` before the `setText()` call but must be resetted to the base type (`string | null`) after `setText()` has been called because this function may have side-effects and therefor TypeScript can no longer assume that `text` is still `null`. The `if (text)` null-check must then change the type to `string` and then the `toLowerCase` method call must not generate an error.

**Actual behavior:**

In the `if (text)` line TypeScript thinks that `text` has the type `null` because it sees no other assignment than then initial one and ignores the possible side-effect of the called `setText()` function. The null check eliminates this type so it becomes type `never` and this ends in the following error message:

    test.ts(9,22): error TS2339: Property 'toLowerCase' does not exist on type 'never'.

**Comment by ahejlsberg** [comment]
See #9998.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
